### PR TITLE
[APB-1130][ZF,AO] validate continue url and display continue button on agent's account page

### PR DESCRIPTION
**Pull request by arturopala**

### CODE REVIEW

File: app/uk/gov/hmrc/agentservicesaccount/services/HostnameWhiteListService.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import uk.gov.hmrc.agentservicesaccount.AppConfig
+import uk.gov.hmrc.agentservicesaccount.connectors.SsoConnector
+import uk.gov.hmrc.play.binders.ContinueUrl
+import uk.gov.hmrc.play.http.HeaderCarrier
+
+import scala.concurrent.Future
+import scala.util.Try
+
+@Singleton
+class HostnameWhiteListService @Inject()(appConfig: AppConfig, ssoConnector: SsoConnector) {
```
Comment (johnrevill) [inline app/uk/gov/hmrc/agentservicesaccount/services/HostnameWhiteListService.scala]: I can see scoverage covers this service, but there's no unit tests specifically covering it. Do we need any?

File: test/uk/gov/hmrc/agentservicesaccount/controllers/AgentServicesControllerSpec.scala [hunk 2]
```diff
@@ -16,66 +16,160 @@
...
+    "do not fail without continue url parameter" in {
+      val controller = new AgentServicesController(messagesApi, authActions, continueUrlActions, externalUrls)
+      val response = controller.root().apply(FakeRequest("GET", "/"))
+      status(response) shouldBe OK
+      contentType(response).get should not include "<button class=\"btn button\" type=\"continue\""
+    }
+
+    "support relative continue url parameter" in {
+      val controller = new AgentServicesController(messagesApi, authActions, continueUrlActions, externalUrls)
+      val response = controller.root().apply(FakeRequest("GET", "/?continue=/foo"))
```
Comment (johnrevill) [inline test/uk/gov/hmrc/agentservicesaccount/controllers/AgentServicesControllerSpec.scala]: Trivial: URLEncoder.encode(url, "UTF-8") is used on line 143 and in some other tests below, but not here even though there's a '/' in the parameter value. 
Same deal with line 133.

File: app/uk/gov/hmrc/agentservicesaccount/controllers/ContinueUrlActions.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      val continueUrl = request.getQueryString("continue").fold[Future[Option[ContinueUrl]]](
+        Future.successful(None)
+      ) { continueParam =>
+        Try(ContinueUrl(continueParam)) match {
+          case Success(url) =>
+            isRelativeOrAbsoluteWhiteListed(url).
+              map(if (_) Some(url) else None).
+              recover {
+                case NonFatal(e) =>
+                  Logger.warn(s"Check for whitelisted hostname failed", e)
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentservicesaccount/controllers/ContinueUrlActions.scala]: we check only against `Host` part of URL

File: app/uk/gov/hmrc/agentservicesaccount/services/HostnameWhiteListService.scala (hunk 1 above)
Comment (arturopala) [inline app/uk/gov/hmrc/agentservicesaccount/services/HostnameWhiteListService.scala]: same here: we check only against `Host` part of URL

Comment (arturopala) [inline app/uk/gov/hmrc/agentservicesaccount/services/HostnameWhiteListService.scala]: I find this advice practical: "A **reasonable combination** of unit and integration tests ensures that every single unit works correctly, independently from others, and that all these units play nicely when integrated, giving us a high level of confidence that the whole system works as expected. " from [how-to-write-testable-code](https://www.toptal.com/qa/how-to-write-testable-code-and-why-it-matters)

File: test/uk/gov/hmrc/agentservicesaccount/controllers/AgentServicesControllerSpec.scala (hunk 2 above)
Comment (arturopala) [inline test/uk/gov/hmrc/agentservicesaccount/controllers/AgentServicesControllerSpec.scala]: we use url encoding only for test urls having query part, `/` doesn't need encoding

File: app/uk/gov/hmrc/agentservicesaccount/connectors/GGRegistrationFEHttpMonitor.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.agentservicesaccount.connectors
+
+import uk.gov.hmrc.play.events.monitoring.HttpErrorMonitor
+
+trait GGRegistrationFEHttpMonitor extends HttpErrorMonitor {
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentservicesaccount/connectors/GGRegistrationFEHttpMonitor.scala]: Good spot, I will change it

File: app/uk/gov/hmrc/agentservicesaccount/frontendAppConfig.scala [hunk 5]
```diff
@@ -24,17 +26,23 @@ trait AppConfig {
...
   private val contactFormServiceIdentifier = "MyService"
 
-  override lazy val analyticsToken = loadConfig(s"google-analytics.token")
-  override lazy val analyticsHost = loadConfig(s"google-analytics.host")
+  override lazy val analyticsToken = loadConfig("google-analytics.token")
+  override lazy val analyticsHost = loadConfig("google-analytics.host")
   override lazy val reportAProblemPartialUrl = s"$contactHost/contact/problem_reports_ajax?service=$contactFormServiceIdentifier"
   override lazy val reportAProblemNonJSUrl = s"$contactHost/contact/problem_reports_nonjs?service=$contactFormServiceIdentifier"
+
+  override lazy val domainWhiteList = runModeConfiguration.getStringList("continueUrl.domainWhiteList").getOrElse(Collections.emptyList()).toSet
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentservicesaccount/frontendAppConfig.scala]: Of course, we should not copy-paste out-of-hand.

File: app/uk/gov/hmrc/agentservicesaccount/frontendAppConfig.scala [hunk 6]
```diff
@@ -24,17 +24,21 @@ trait AppConfig {
...
   private val contactFormServiceIdentifier = "MyService"
 
-  override lazy val analyticsToken = loadConfig(s"google-analytics.token")
-  override lazy val analyticsHost = loadConfig(s"google-analytics.host")
+  override lazy val analyticsToken = loadConfig("google-analytics.token")
+  override lazy val analyticsHost = loadConfig("google-analytics.host")
   override lazy val reportAProblemPartialUrl = s"$contactHost/contact/problem_reports_ajax?service=$contactFormServiceIdentifier"
   override lazy val reportAProblemNonJSUrl = s"$contactHost/contact/problem_reports_nonjs?service=$contactFormServiceIdentifier"
+
+  override lazy val domainWhiteList = runModeConfiguration.getStringSeq("continueUrl.domainWhiteList")
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentservicesaccount/frontendAppConfig.scala]: This now returns ```Option[Seq[String]]```. I think we should use get Set[String] from it?

Comment (arturopala) [inline app/uk/gov/hmrc/agentservicesaccount/frontendAppConfig.scala]: thanks, fixed

**Pull request merged**